src/containers/methodCallContainer.js

```diff
@@ -1,42 +1,101 @@
 import buildProvider from '../helpers/buildProvider';
+import {
+  fetchOrParseAbi,
+  getFilteredMethods,
+  getArgumentsFromMethodId,
+  getContractFriendlyArguments,
+} from '../helpers/contracts';
 import { navigate, useParams } from '@reach/router';
 
+const CONTRACT_FUNCTION_METHOD = 'eth_call';
+
 const MethodCallContainer = () => {
   const params = useParams();
-  const { codeSampleVisible, toggleSampleCode } = useContext(AppContext);
+  const { codeSampleVisible, toggleSampleCode, abi, setAbi } = useContext(
+    AppContext
+  );
   const { addToLog } = useContext(LogContext);
   const logItem = useCallback(addToLog, []);
+
   const {
     web3URL = '',
     web3Lib = '',
     currentMethod = '',
     formArgs = '',
   } = params;
-  const argumentList = formArgs.split('/');
+
   const web3Method = Web3RpcCalls[currentMethod] || {};
   const { description, disabled } = web3Method || {};
-  const { args, exec } = web3Method[web3Lib] || {};
-  const setArgumentList = (val, index) => {
-    const argsCopy = [...argumentList];
-    argsCopy[index] = val;
-    // bad pattern?
-    let joinedArgs = argsCopy.join('/');
+  const { args: initialFormInputs, exec } = web3Method[web3Lib] || {};
+
+  const [formInputs, setFormInputs] = useState([]);
+  const [argumentList, setArgumentList] = useState([]);
+
+  const updateURL = (val, index) => {
+    const argsList = formArgs.split('/');
+    argsList[index] = val;
+    let joinedArgs = argsList.join('/');
     let url = `/${web3URL}/${web3Lib}/`;
     if (currentMethod) url += `${currentMethod}/`;
-    if (args.length > 0) url += `${joinedArgs}`;
+    if (formInputs.length > 0) url += `${joinedArgs}`;
     navigate(url);
   };
-  const runRequest = (args) => {
+
+  const onUpdateContractMethod = () => {
+    const methodId = argumentList[2];
+    if (!methodId) return;
+    const newFormInputs = getArgumentsFromMethodId(methodId);
+    if (newFormInputs)
+      setFormInputs([
+        ...formInputs.slice(0, 3), // Discard existing method-specific inputs
+        ...newFormInputs,
+      ]);
+    else setFormInputs([...formInputs.slice(0, 3)]);
+  };
+
+  const onUpdateArguments = async (val, index) => {
+    if (currentMethod === CONTRACT_FUNCTION_METHOD) {
+      if (index === 1) {
+        // Prevent updating URL if ABI error
+        const { error } = await fetchOrParseAbi(val);
+        if (error)
+          return logItem({
+            method: 'error',
+            data: ['🚨 Error:', error],
+          });
+        return updateURL(btoa(val), index);
+      }
+    }
+    updateURL(val, index);
+  };
+
+  const onUpdateAbi = () => {
+    const filteredMethods = getFilteredMethods(abi);
+    const formInputsCopy = formInputs;
+    formInputsCopy[2] = {
+      ...formInputs[2],
+      dropdownOptions: filteredMethods,
+      disabled: abi.length === 1,
+    };
+    setFormInputs(formInputsCopy);
+    if (abi.length === 1) updateURL(filteredMethods[0].value, 2);
+  };
```

Comment:
Same thing here, we can lift this out of here into some other file, pass in args and return values to be updated in the `useEffect` call.

Reply:
Done